src/nerdbank-streams/src/Deferred.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * A TaskCompletionSource-like class that allows promises to be resolved or rejected whenever.
+ */
+export class Deferred<T> {
+    public readonly promise: Promise<T>;
+    private resolvePromise: (value?: T | PromiseLike<T>) => void;
+    private rejectPromise: (reason?: any) => void;
+    private _isResolved: boolean = false;
+    private _isRejected: boolean = false;
+    private _error: any;
+
+    constructor(public state?: any) {
```

Comment:
I am a little surprised you didn't make this another type parameter

Reply:
`state` is only used in one place. But this class is used all over. Requiring everyone to give a type argument for something they don't use seems tedious. Also, this resembles `TaskCompletionSource<T>`, which just takes `object` for its state rather than trying to make it type-safe.